### DISCUSSION TITLE
Method is not exposed to Udon: 'inputField.text', 'dropdown.value'

### DISCUSSION THREAD

**Discussion by baythebassist** [wrchat/Woodon#25] [Q&A] [answered]
## 문제 설명
- TMP 를 사용할 수가 없어서 Unity 의 legacy UI component 로 바꿔서 쓰고 있습니다.
- TMP 를 쓰려면 어떻게 해야하는 걸까요?
- 참고로 TMP Essential Resources, Examples and Extras 은 import 한 상태입니다!

- [TextMesh Pro docs](https://creators.vrchat.com/worlds/components/textmeshpro/) 보면 exposed to Udon 이라고 하는데, 에러가 뜹니다...
![image](https://github.com/user-attachments/assets/aea30953-6a11-416e-ae47-46b6284dc64d)

- [Class Exposure Tree](https://udonsharp.docs.vrchat.com/class-exposure-tree/) 에 TMP_Dropdown 이랑 TMP_InputField 가 없습니다.

![image](https://github.com/user-attachments/assets/87b0b230-b24e-47c5-ab64-0310874453e8)

## 버전 정보
- Unity version : 2022.3.22
- VRC SDK : 3.6.0
- MUdons release version : 0.0.36

## 에러 메시지
```
[UdonSharp] Packages/com.wakvrc.m-udons/Runtime/CustomValue/MString/Scripts/UI/UIMString.cs(31,4): Method is not exposed to Udon: 'inputField.text'
[UdonSharp] Packages/com.wakvrc.m-udons/Runtime/MUI/KoreanKeyboard/KoreanKeyboard.cs(57,6): Method is not exposed to Udon: 'syncInputField.text'
[UdonSharp] Packages/com.wakvrc.m-udons/Runtime/TurnBase/MDraw/Scripts/UI/UIDrawTeamSetter.cs(19,56): Method is not exposed to Udon: 'dropdown.value'
```

## 해결 방법
> 그런데 이제 이 방법으로 해결하면 안될 것 같은... 🤔

UIMString.cs
`[SerializeField] private TMP_InputField inputField;` → `[SerializeField] private InputField inputField;` 

KoreanKeyboard.cs
`[SerializeField] private TMP_InputField syncInputField;` → `[SerializeField] private InputField syncInputField;`

UIDrawTeamSetter.cs
`[SerializeField] private TMP_Dropdown dropdown;` → `[SerializeField] private Dropdown dropdown;`

----

- 스크립트 이렇게 변경하고 UI component 도 legacy 로 다 세팅해주면 되긴 합니다...!
- 그런데 이제 슬슬 TMP 를 쓰고 싶어서... 디스커션 남깁니당
- 따로 Udon에 exposed 될 함수를 만들어줘야 하는걸까요?

**Comment by mascari4615** [comment, marked as answer]
## 해결 방법

[VRChat SDK 3.6.1](https://creators.vrchat.com/releases/release-3-6-1) 버전부터 TMP를 지원합니다.
크게 문제가 없다면, 3.6.1 이상으로 버전 업 해주시면 해결될 것 같아요.

## 조치

MUdons **프로젝트** 자체는 [vpm-manifest.json](https://github.com/WakVRC/MUdons/blob/main/Packages/vpm-manifest.json)으로 가장 최신 버전의 VRChat SDK로 유지하고 있는데,
MUdons **유니티 패키지**는 따로 디펜던시를 적어두지 않아서, 이걸 써서 요구되는 VRChat SDK 버전을 강제할 필요가 있겠네요.

혹은 Readme에 요구되는 VRChat SDK를 적어둬야겠습니다.

**Reply by mascari4615** [reply]
\+ 지금 시점에서 MUdons 유니티 패키지는 MUdons 프로젝트와 마찬가지로 VRChat SDK 3.7.1 버전을 사용하고 있습니다.

지금 시점에서는 `3.7.2-beta.3`이 **가장** 최신이긴한데, 크게 바뀌는 부분이 없어 `3.7.1` 버전으로 유지하고 있습니다.

**Reply by baythebassist** [reply]
오 `3.7.1` 로 버전 업했더니 잘 되네요! 감사합니다 👍
가이드 문서 있으면 좋을 것 같아서 wiki에 [Getting Started](https://github.com/WakVRC/MUdons/wiki/Getting-Started) 페이지 만들어놨는데, 나중에 여기에도 정리해놔야겠네요...!
